[RFC] Wifi socket support

ryandoherty opened the pull request:
Initial PR for socket support for our app. This is requesting to merge into ios-support so only the changes for sockets are shown. Do not merge!

Important files are:

* socketcomm.py
* socketconnection.py

Other changes are mostly just renaming `port` to `device`.

Socket communication is nearly identical to serial communication. Separate threads read and write data to the socket and put/pull data from shared queues. Socket connection continually watches for data from the socket unless a shared event object is false, then it returns and ends.

Todo:
* tests

brentpicasso commented on autosportlabs/comms/commsfactory.py:
```diff
@@ -8,13 +8,17 @@
+        from autosportlabs.comms.socket.socketconnection import SocketConnection
+        from autosportlabs.comms.socket.socketcomm import SocketComm
+        return SocketComm(SocketConnection(), device)
+        # from autosportlabs.comms.comms import Comms
```
commented out code

ryandoherty replied:
Ah, yeah I did this to use wifi on my mac. Will uncomment and remove socket stuff from this before I merge in.

brentpicasso commented on autosportlabs/comms/socket/socketcomm.py:
```diff
@@ -1,46 +1,158 @@
 COMMAND_KEEP_ALIVE = 'PING'
 
+
+def connection_thread_message_reader(rx_queue, connection, should_run):
```
Recommend making these functions part of the SocketComm class, maybe mark them as private methods (_)

ryandoherty replied:
👍

brentpicasso commented:
please add class and method comments per contributing doc using http://docutils.sourceforge.net/rst.html format

brentpicasso commented on autosportlabs/comms/socket/socketcomm.py:
```diff
@@ -1,46 +1,158 @@
     def start_connection_process(self):
-        pass
+        rx_queue = Queue.Queue()
+        tx_queue = Queue.Queue(5)
```
extract magic number; applies throughout

brentpicasso commented on autosportlabs/comms/socket/socketcomm.py:
```diff
@@ -1,46 +1,158 @@
+            if msg:
+                rx_queue.put(msg)
+        except:
+            Logger.error('SocketComm: Exception in connection_process_message_reader')
```
looks like an exception in either thread will cause both to exit. intentional?

Also, do we want to exit on all exceptions? Quite possibly, but the connection might throw an exception that could be  benign enough to not warrant closing everything down possibly. thoughts?

ryandoherty replied:
I'm not sure, this is the same code as the serial comms, so I figured whatever behavior it did was best to follow since code that uses serial comm also uses socket comm.

brentpicasso replied:
ok

ryandoherty replied:
From what I can tell, the RCP API code will restart connections (BT, USB and WiFi) if they fail or stop responding. Right now I don't know what types of exceptions would occur that would be safe to continue to process data with. So I think we're ok with leaving this behavior in as it's the same as the BT & USB behavior. 

Once I have a WiFi module to do some real testing against I can see what types of errors may occur that would warrant keeping the thread running.

brentpicasso commented on autosportlabs/comms/socket/socketconnection.py:
```diff
@@ -1,31 +1,79 @@
+
+                msg += data
+
+                if msg[-2:] == '\r\n':
```
what if we only get one character? will -2 generate an array index out of bound exception?

ryandoherty replied:
According to python docs, out of range indexes when slicing results in a blank characters and no exceptions thrown, so we're safe.

brentpicasso replied:
fantastical!

brentpicasso commented on autosportlabs/racecapture/api/rcpapi.py:
```diff
@@ -666,28 +666,29 @@ def on_ver_win(value):
-                    if last_known_port:
-                        Logger.info('RCPAPI: trying last known port first: {}'.format(last_known_port))
+                    devices = comms.get_available_devices()
+                    # Backwards compatibility when fetching port, convert to device
```
i think it would be safe to convert this preferences key to the name 'device' . the impact is that the value won't be set, and it'll have to do an initial search again.

ryandoherty replied:
👍

brentpicasso commented on autosportlabs/racecapture/databus/databus.py:
```diff
@@ -190,11 +190,11 @@ def start(self, data_bus, rc_api):
-            t = Thread(target=self.sample_worker)
-            t.start()
-            self._sample_thread = t
+        # if not is_mobile_platform():
```
left over commented code?

ryandoherty replied:
Not sure, will check.

ryandoherty replied:
I commented this out so the sample worker wouldn't poll the mock wifi service I wrote. Un-commented.

brentpicasso commented:
all of this is exciting to see.

ryandoherty commented:
Ok @brentpicasso I have added comments, fixed the things that needed to be fixed and swapped the polling support to not use is_mobile() and instead query the comms object (for now, long term we need the RC firmware API to expose this).

brentpicasso commented on autosportlabs/racecapture/views/configuration/rcp/scriptview.kv:
```diff
@@ -13,28 +13,28 @@
             ScrollContainer:
                 size_hint_y: 0.75
                 do_scroll_x:False
-                id: luaScriptScrollView
```
what's going on here? merge stuff that's out of sync with other things?

brentpicasso commented:
Looked; your changes look good but it seems there's extra stuff here from previous changes (around script window, etc). Maybe not 100% up to sync with branch?

ryandoherty commented:
Oops yeah it's because i merged master into this branch but not the branch
this PR was setup to land in. Will fix.

-Ryan
On May 11, 2016 4:12 PM, "Brent Picasso" <notifications@github.com> wrote:

Looked; your changes look good but it seems there's extra stuff here from
previous changes (around script window, etc). Maybe not 100% up to sync
with branch?

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub
<https://github.com/autosportlabs/RaceCapture_App/pull/977#issuecomment-218617278>

ryandoherty commented:
Closing, will merge manually into master.
